Mono WASM PInvoke Table generation does not handle file scoped types

### Description

File scoped types are named like `<File_Name_Without_Periods>Hash__ClassName`. For `UnmanagedCallersOnly` methods in file-scoped types, Mono tries to generate a method in C, but it doesn't escape the name. As a result, you get errors like the following:

```
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error : In file included from /root/helix/work/correlation/build/microsoft.netcore.app.runtime.browser-wasm/runtimes/browser-wasm/native/src/pinvoke.c:15: [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error : /root/helix/work/workitem/e/wasm_build/obj/wasm/for-build/pinvoke-table.h:212:64: error: expected ';' after top level declarator [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error : int wasm_native_to_interp_System_Runtime_InteropServices_Tests_<System_Runtime_InteropServices_Tests_ComVariantMarshallerTests_IGeneratedComInterface>F3F2786D8F8679299059CA9E93026F23079CBD4B40DBBE7EBFBCA0FE1F6116567__InterfaceImplementation_ABI_Method (int arg0) {  [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error :                                                                ^ [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error :                                                                ; [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error : /root/helix/work/workitem/e/wasm_build/obj/wasm/for-build/pinvoke-table.h:220:3: error: use of undeclared identifier 'WasmInterpEntrySig_5' [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error : ((WasmInterpEntrySig_5)wasm_native_to_interp_ftndescs [5].func) ((int*)&res, (int*)&arg0, (int*)&arg1, wasm_native_to_interp_ftndescs [5].arg); [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error :   ^ [/root/helix/work/workitem/e/publish/ProxyProjectForAOTOnHelix.proj]
/root/helix/work/correlation/build/wasm/WasmApp.Native.targets(367,5): error : /root/helix/work/workitem/e/wasm_build/obj/wasm/for-build/pinvoke-table.h:269:420: error: use of undeclared identifier 'System_Runtime_InteropServices_Tests_ComVariantMarshallerTests_IGeneratedComInterface' 
```

### Reproduction Steps

Build https://github.com/dotnet/runtime/pull/93635 in the LibraryTests_EAT configuration.

### Expected behavior

Build succeeds

### Actual behavior

Build fails with above mentioned error.

### Regression?

Unknown

### Known Workarounds

_No response_

### Configuration

Running in a PR into this repo: https://github.com/dotnet/runtime/pull/93635

### Other information

Helix log for the failing test:

https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-93635-merge-9667f32037a34a4fa0/normal-System.Runtime.InteropServices.Tests/1/console.147620a6.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-93635-merge-9667f32037a34a4fa0/normal-System.Runtime.InteropServices.Tests/1/console.147620a6.log?helixlogtype=result)

--- a/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
+++ b/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
@@ -376,8 +376,8 @@ namespace Wasm.Build.Tests
         }
 
         [Theory]
-        [BuildAndRun(host: RunHost.None)]
-        public void UnmanagedCallback_InFileType_Compiles(BuildArgs buildArgs, string id)
+        [BuildAndRun(host: RunHost.Chrome)]
+        public void UnmanagedCallback_InFileType(BuildArgs buildArgs, string id)
         {
             string code =
                 """
@@ -399,13 +399,16 @@ namespace Wasm.Build.Tests
                 }
                 """;
 
-            (_, string output) = BuildForVariadicFunctionTests(
+            (buildArgs, string output) = BuildForVariadicFunctionTests(
                 code,
                 buildArgs with { ProjectName = $"cb_filetype_{buildArgs.Config}" },
                 id
             );
 
             Assert.DoesNotMatch(".*(warning|error).*>[A-Z0-9]+__Foo", output);
+    
+            output = RunAndTestWasmApp(buildArgs, buildDir: _projectDir, expectedExitCode: 42, host: host, id: id);
+            Assert.Contains("Main running", output);
         }
 
         [Theory]

--- a/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
+++ b/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
@@ -377,6 +377,39 @@ namespace Wasm.Build.Tests
 
         [Theory]
         [BuildAndRun(host: RunHost.None)]
+        public void UnmanagedCallback_InFileType_Compiles(BuildArgs buildArgs, string id)
+        {
+            string code =
+                """
+                using System;
+                using System.Runtime.InteropServices;
+                public class Test
+                {
+                    public static int Main()
+                    {
+                        Console.WriteLine("Main running");
+                        return 42;
+                    }
+                }
+
+                file class Foo
+                {
+                    [UnmanagedCallersOnly]
+                    public unsafe static extern void SomeFunction1(delegate* unmanaged<int> callback);
+                }
+                """;
+
+            (_, string output) = BuildForVariadicFunctionTests(
+                code,
+                buildArgs with { ProjectName = $"cb_filetype_{buildArgs.Config}" },
+                id
+            );
+
+            Assert.DoesNotMatch(".*(warning|error).*>[A-Z0-9]+__Foo", output);
+        }
+
+        [Theory]
+        [BuildAndRun(host: RunHost.None)]
         public void IcallWithOverloadedParametersAndEnum(BuildArgs buildArgs, string id)
         {
             // Build a library containing icalls with overloaded parameters.

--- a/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
+++ b/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
@@ -377,7 +377,7 @@ namespace Wasm.Build.Tests
 
         [Theory]
         [BuildAndRun(host: RunHost.Chrome)]
-        public void UnmanagedCallback_InFileType(BuildArgs buildArgs, string id)
+        public void UnmanagedCallback_InFileType(BuildArgs buildArgs, RunHost host, string id)
         {
             string code =
                 """

--- a/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
+++ b/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs
@@ -395,7 +395,7 @@ namespace Wasm.Build.Tests
                 file class Foo
                 {
                     [UnmanagedCallersOnly]
-                    public unsafe static extern void SomeFunction1(delegate* unmanaged<int> callback);
+                    public unsafe static extern void SomeFunction1(int i);
                 }
                 """;
 

--- a/src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs
+++ b/src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs
@@ -35,7 +35,7 @@ public class ManagedToNativeGenerator : Task
     [Output]
     public string[]? FileWrites { get; private set; }
 
-    private static readonly char[] s_charsToReplace = new[] { '.', '-', '+' };
+    private static readonly char[] s_charsToReplace = new[] { '.', '-', '+', '<', '>' };
 
     // Avoid sharing this cache with all the invocations of this task throughout the build
     private readonly Dictionary<string, string> _symbolNameFixups = new();

--- a/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs
+++ b/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs
@@ -308,7 +308,7 @@ internal sealed class PInvokeTableGenerator
 
             string module_symbol = _fixupSymbolName(method.DeclaringType!.Module!.Assembly!.GetName()!.Name!);
             uint token = (uint)method.MetadataToken;
-            string class_name = method.DeclaringType.Name;
+            string class_name = _fixupSymbolName(method.DeclaringType.Name);
             string method_name = method.Name;
             string entry_name = $"wasm_native_to_interp_{module_symbol}_{class_name}_{method_name}";
             if (callbackNames.Contains(entry_name))

--- a/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs
+++ b/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs
@@ -373,10 +373,10 @@ internal sealed class PInvokeTableGenerator
         foreach (var cb in callbacks)
         {
             var method = cb.Method;
-            string module_symbol = method.DeclaringType!.Module!.Assembly!.GetName()!.Name!;
+            string module_symbol = _fixupSymbolName(method.DeclaringType!.Module!.Assembly!.GetName()!.Name!);
             string class_name = method.DeclaringType.Name;
             string method_name = method.Name;
-            w.WriteLine(_fixupSymbolName($"\"{module_symbol}_{class_name}_{method_name}\","));
+            w.WriteLine($"\"{module_symbol}_{class_name}_{method_name}\",");
         }
         w.WriteLine("};");
     }

--- a/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs
+++ b/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs
@@ -306,11 +306,11 @@ internal sealed class PInvokeTableGenerator
 
             bool is_void = method.ReturnType.Name == "Void";
 
-            string module_symbol = _fixupSymbolName(method.DeclaringType!.Module!.Assembly!.GetName()!.Name!);
+            string module_symbol = method.DeclaringType!.Module!.Assembly!.GetName()!.Name!;
             uint token = (uint)method.MetadataToken;
-            string class_name = _fixupSymbolName(method.DeclaringType.Name);
+            string class_name = method.DeclaringType.Name;
             string method_name = method.Name;
-            string entry_name = $"wasm_native_to_interp_{module_symbol}_{class_name}_{method_name}";
+            string entry_name = _fixupSymbolName($"wasm_native_to_interp_{module_symbol}_{class_name}_{method_name}");
             if (callbackNames.Contains(entry_name))
             {
                 Error($"Two callbacks with the same name '{method_name}' are not supported.");
@@ -373,10 +373,10 @@ internal sealed class PInvokeTableGenerator
         foreach (var cb in callbacks)
         {
             var method = cb.Method;
-            string module_symbol = _fixupSymbolName(method.DeclaringType!.Module!.Assembly!.GetName()!.Name!);
+            string module_symbol = method.DeclaringType!.Module!.Assembly!.GetName()!.Name!;
             string class_name = method.DeclaringType.Name;
             string method_name = method.Name;
-            w.WriteLine($"\"{module_symbol}_{class_name}_{method_name}\",");
+            w.WriteLine(_fixupSymbolName($"\"{module_symbol}_{class_name}_{method_name}\","));
         }
         w.WriteLine("};");
     }

